### ISSUE TITLE
I can't reproduce results after restarting kernel in Jupyter with LightGBM 2.3.1

### ISSUE THREAD

**Issue by santigc6** [microsoft/LightGBM#5243]
Hi, I am trying to reproduce LightGBM results over several runs. I am using the following parameters:

`{'boosting_type': 'gbdt',
 'objective': 'binary',
 'metric': 'auc',
 'subsample_freq': 1,
 'is_unbalance': False,
 'verbose': -1,
 'seed': 42,
 'max_bin': 255,
 'learning_rate': 0.05,
 'subsample': 1.0,
 'colsample_bytree': 0.75,
 'deterministic': True,
 'monotone_constraints': array([ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0, -1,  0,  0,  0,  0,  0,  0,
         0,  0,  0,  0,  0,  0,  0,  0]),
 'num_leaves': 127,
 'max_depth': 6,
 'min_sum_hessian_in_leaf': 0,
 'min_split_gain': 0,
 'reg_alpha': 0,
 'reg_lambda': 0,
 'num_threads': 1,
 'force_row_wise': True,
 'force_col_wise': True}`

The rest of the parameters are set to their default values. And the thing is that I if I run a CV with a new model without restarting Jupyter's kernel, I get exactly the same results. But, once I restart the kernel, I don't get the same results (of the executions before restarting the kernel) anymore. i.e. with the same kernel I can reproduce the results I get using that kernel, but once I restart the kernel the results change.

![image](https://user-images.githubusercontent.com/22822055/170473186-be95de3b-7f71-4d0d-8b85-8c7564649863.png)
 
And once I restart the kernel

![image](https://user-images.githubusercontent.com/22822055/170473339-6936a0b9-3942-40b7-b83d-3a714a62db80.png)

ipykernel==5.3.4
ipympl==0.5.8
ipython==7.16.1
ipython-genutils==0.2.0
ipython-sql==0.4.0
ipywidgets==7.5.1
jupyter-client==6.1.6
jupyter-core==4.6.3
jupyter-tensorboard==0.2.0
jupyterhub==0.9.2
jupyterlab==2.2.10
jupyterlab-git==0.23.3
jupyterlab-launcher==0.13.1
jupyterlab-pygments==0.1.2
jupyterlab-server==1.2.0
jupytext==1.10.0
lightgbm==2.3.1
nbclient==0.5.9
nbconvert==6.0.7
nbdime==2.1.0
nbformat==5.0.8
nbresuse==0.4.0
nbsphinx==0.8.1
numpy==1.19.5

**Comment by StrikerRUS** [comment]
Hey @santigc6 , thanks for your question!

You are using VERY old LightGBM version. That version even doesn't contain `deterministic` parameter which you're passing in your code. Please consider updating LightGBM to the latest version.

Also, without any reproducible example we won't be able to help you.

**Comment by santigc6** [comment]
Okay, I have also tried with versions 3.3.1 and 3.3.2. It works just fine as long as I don't use monotone_constraints. But once I use monotone_constraints, result changes when I restart the kernel.

Problem is I have everything wrapped and I can't send you the code (confidential reasons). But I will keep trying to solve it and if I manage to do it, I will share the solution. Thanks!

**Comment by santigc6** [comment]
Hi, I have managed to finally solve the problem. The issue was related to column order with categorical features when generating the lightgbm.Dataset. You have to use categorical_feature parameter to fix it. So in short, once you set `{"deterministic": True, "force_col_wise": True}` and use `categorical_feature` parameter to generate the dataset it works just fine. I have also set `{"seed": 42}`, but there is probably no need to do it.

**Comment by jmoralez** [comment]
Hi @santigc6. Just to clarify, using `categorical_feature='auto'` was the source of instability?

**Comment by santigc6** [comment]
> Hi @santigc6. Just to clarify, using `categorical_feature='auto'` was the source of instability?

In my case it was! `If 'auto' and data is pandas DataFrame, pandas unordered categorical columns are used.`
